File: check_dashboard.go
/*
* 检测codis中的dashboard进程
*     如果该进程挂掉，则重启它
*         因为zookeeper的缘故，在重启之前需要将zk中的dashboard信息delete掉
*
* 使用前，
*     配置zk连接信息与dashboard命令
*     你可能需要执行 go get github.com/samuel/go-zookeeper/zk
* github.com/chenshangwei/check_codis_dashboard
*/

package main

import (
   "fmt"
   "time"
   "log"
   "os"
   "os/exec"
   "net/http"
   s "strings"
   "github.com/samuel/go-zookeeper/zk"
)

/*config*/
var zk_ip  = []string{"192.168.1.200"}
var addr string = "192.168.1.200:18087"
var zk_dashboard = "/zk/codis/db_120ask/dashboard"
var dashboard_command string = "/home/go/src/github.com/CodisLabs/codis/bin/codis-config"
var dashboard_args string = "-c /home/go/src/github.com/CodisLabs/codis/config.ini -L /home/go/src/github.com/CodisLabs/codis/logs/dashboard.log dashboard --addr "+addr  //最后不加&
var logFile string = "/root/tools/check_dashboard.log"


//http请求，判断dashboard是否存活
func check() bool{
   url := "http://"+addr+"/api/overview"
   //url := "http://www.baidu.com"
   if resp,err := http.Get(url);err != nil {
      return false
   } else {
      resp.Body.Close()
      return true
   }
}
//记录日志
func writeLog(msg string){
   logfile,err := os.OpenFile(logFile,os.O_RDWR|os.O_CREATE|os.O_APPEND,0666);
   if err!=nil {
      panic(err)
   }
   defer logfile.Close();
   logger := log.New(logfile,"\r\n",log.Ldate|log.Ltime|log.Llongfile);
   logger.Fatal(msg);
}
//启动dashboard
func runDashboard(){
   fmt.Println("exec : " + dashboard_command)
   var argArray []string
   argArray = s.Split(dashboard_args," ")
   
   cmd := exec.Command(dashboard_command,argArray...)
   if err := cmd.Start();err != nil {
      writeLog("[fail]start dashboard")
   }else{
      writeLog("[success]start dashboard")
   }
}
//删掉zk中的dashboard信息
func delDashboard() {
   fmt.Println("delete : " + zk_dashboard)
   if c,_, err := zk.Connect(zk_ip, time.Second*2);err != nil {
      writeLog("[fail]connect zk")
   }else{
      d,_,_ := c.Get(zk_dashboard)
      fmt.Printf("%s",d)
      if len(d)>0 {
         err := c.Delete(zk_dashboard,-1)
         if err != nil {
            writeLog("[fail]delete zk_dashboard")
         }else{
            writeLog("[success]delete zk_dashboard")
         }
      }
   }
}
func main() {
   if check() {
      fmt.Println("zk dashboard is ok")
      //os.Exit(0)
   }
   delDashboard()
   runDashboard()
}

File: check_dashboard.py
#!/usr/bin/env python
#coding:utf-8
# pip install kazoo==2.0   #kazoo2.21使用了six模块在python2.7下有问题，故使用kazoo2.0版本
'''
    pip安装模块警告InsecurePlatformWarning: A true SSLContext object is not available.
    解决方法见：http://blog.csdn.net/henulwj/article/details/48131393/
    
         首先安装Python-devel libffi-devel openssl-devel
            yum install python-devel libffi-devel openssl-devel
         之后在安装pyopenssl ndg-httpsclient pyasn1
            pip install pyopenssl ndg-httpsclient pyasn1
'''
# https://github.com/chenshangwei/check_codis_dashboard/

'''
   检测codis中的dashboard进程
       如果该进程挂掉，则重启它
           因为zookeeper的缘故，在重启之前需要将zk中的dashboard信息delete掉
   python check_dashboard.py
   使用前，
     配置zk连接信息与dashboard命令
'''
import sys
import logging
import urllib2
import subprocess
import time
from kazoo.client import KazooClient

zk_hosts = "192.168.1.200:2181"
zk_dashboard = "/zk/codis/db_120ask/dashboard"
dashboard_addr = "192.168.1.200:18087"

dashboard_command = "/home/go/src/github.com/CodisLabs/codis/bin/codis-config -c /home/go/src/github.com/CodisLabs/codis/config.ini -L /home/go/src/github.com/CodisLabs/codis/logs/dashboard.log dashboard --addr " + str(dashboard_addr) + " &"

codisha_command = "/home/go/src/github.com/ngaut/codis-ha/codis-ha -codis-config=" + str(dashboard_addr) + " -productName=120ask &"

logging.basicConfig(level=logging.INFO,
                format='%(asctime)s %(message)s',
                datefmt='%Y-%b-%d %H:%M:%S',
                filename='/root/tools/check_dashboard.log', #日志文件
                filemode='ab')

def delDashboard():
   zk = KazooClient(hosts=zk_hosts)
   zk.start()
   if zk.exists(zk_dashboard):
      if zk.delete(zk_dashboard):
         logging.info("[success] delete %s" % zk_dashboard)     
   zk.stop()

def runDashboard():
   recode = subprocess.call(dashboard_command,shell=True)
   if recode == 0:
      logging.info("[success] start dashboard")
      return True
   else:
      logging.info("[fail] start dashboard")
      return False
def runCodisHA():
   time.sleep(5) #等待dashboard启动
   recode = subprocess.call(codisha_command,shell=True)
   if recode == 0:
      logging.info("[success] start codis-ha")
      return True
   else:
      logging.info("[fail] start codis-ha")
      return False
def check():
   url =  "http://"+str(dashboard_addr)+"/api/overview"
   try:
      request = urllib2.Request(url)
      response = urllib2.urlopen(request)
      return True
   except Exception,e:
      return False

if __name__ == '__main__':
   if check():
      print "dashboard is ok"
      sys.exit(0)
   delDashboard()  #删除zk中的dashboard信息
   runDashboard()  #启动dashboard
   #runCodisHA()   #可选 启动codis-ha
